pinot-core/src/main/java/com/linkedin/pinot/core/io/readerwriter/impl/FixedByteSingleColumnMultiValueReaderWriter.java

```diff
@@ -142,28 +176,42 @@ public void close() {
       dataBuffer.close();
     }
     dataBuffers.clear();
-    headerBuffer.close();
+    for (PinotDataBuffer headerBuffer : headerBuffers) {
+      headerBuffer.close();
+    }
+    headerBuffers.clear();
     headerBuffer = null;
   }
 
-  private int updateHeader(int row, int length) {
-    assert (length < maxNumberOfMultiValuesPerRow);
-    int prevRowStartIndex = 0;
-    int prevRowLength = 0;
-    if (row > 0) {
-      prevRowStartIndex = headerReader.getInt(row - 1, 1);
-      prevRowLength = headerReader.getInt(row - 1, 2);
+  private void writeIntoHeader(int row, int dataWriterIndex, int startIndex, int length) {
+    if (row >= headerBuffers.size() * rowCountPerChunk) {
+      addHeaderBuffers();
     }
+    curHeaderWriter.setInt(getRowInCurrentHeader(row), 0, dataWriterIndex);
+    curHeaderWriter.setInt(getRowInCurrentHeader(row), 1, startIndex);
+    curHeaderWriter.setInt(getRowInCurrentHeader(row), 2, length);
+  }
+
+  private final FixedByteSingleValueMultiColReader getCurrentReader(int row) {
+    return headerReaders.get(row/ rowCountPerChunk);
```

Comment:
Yes, that is why i have it in a separate method. I will add a TODO. Right now, it is not clear yet how the rowsCountPerChunk will be decided. Most likely it will be outside this class, since other classes are also getting re-factored to support multi-tenant realtime